src/query/storage/cache/redis_cache.go

```diff
@@ -41,3 +52,167 @@ func NewRedisCache(redisAddress string, logger *zap.Logger) *RedisCache {
 		logger:       logger,
 	}
 }
+
+// Given a fetch query, converts it into a key for Redis
+func keyEncode(query *storage.FetchQuery) string {
+	res := make([]string, len(query.TagMatchers))
+	for i, m := range query.TagMatchers {
+		res[i] = m.String()
+	}
+	// Sort to guarantee we have the same order every time
+	sort.Strings(res)
+
+	label_key := strings.Join(res, ";")
+	queryRange := int64(query.End.Sub(query.Start).Seconds())
+	// Key becomes {labels}::{endTime}::{duration}
+	return fmt.Sprintf("%s::%d::%d", label_key, query.End.Unix(), queryRange)
+}
+
+// Given a result, encodes it into string (or "" if error)
+// Encodes an empty PromResult to {} to differentiate from ""
+//
+// We only encode the QueryResult portion of the PromResult struct
+// We omit the metadata since we do not use it in computation
+// and also this metadata also isn't accurate since we aren't getting it from M3DB
+func resultEncode(result *storage.PromResult) (string, error) {
+	if len(result.PromResult.Timeseries) == 0 {
+		return EmptyResult, nil
+	}
+	// Custom Marshal function for QueryResults
+	res, err := result.PromResult.Marshal()
+	if err != nil {
+		return "", err
+	}
+	return string(res), nil
+}
+
+// Given a string, decodes it into a result (or nil if failed)
+//
+// Sets the metadata to a blank as we don't use it
+// and we don't actually have metadata from M3DB for this result
+func resultDecode(val []byte) (*storage.PromResult, error) {
+	var result storage.PromResult
+	result.PromResult = &prompb.QueryResult{}
+
+	// Check length first so we don't convert to string every time
+	if len(val) == len([]byte(EmptyResult)) && string(val) == EmptyResult {
+		result.Metadata = block.NewResultMetadata()
+		return &result, nil
+	}
+	// Custom Unmarshal function for QueryResults
+	if err := result.PromResult.Unmarshal(val); err != nil {
+		return nil, err
+	}
+	result.Metadata = block.NewResultMetadata()
+	return &result, nil
+}
+
+// For each entry in {entries}, gives the PromResult or nil if failed in array
+// Array parameter used to handle multiple bucket requests in the future
+func (cache *RedisCache) Get(entries []*storage.FetchQuery) []*storage.PromResult {
+	keys := make([]string, len(entries))
+	results := make([]*storage.PromResult, len(entries))
+
+	for i, b := range entries {
+		keys[i] = keyEncode(b)
+	}
+
+	// Consume as bytes to avoid having to casting to string later on (avoids duplicate memory)
+	responses := make([][]byte, len(entries))
+	if err := cache.client.Do(radix.Cmd(&responses, "MGET", keys...)); err != nil {
```

Comment:
can we add err handling on the error? 
1. retry with backoff
2. emit metrics to read from client with err logging
3. err then fallback to m3 read, do crash coordinator.

Reply:
So to clarify on step 3, right now the code should fallback to reading from M3DB when an error is encountered, but I'm unsure as to what do crash coordinator means

Reply:
sorry, clarify more on the comment

Reply:
Will update in a future PR as this one is already getting a bit big